flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/writer/StreamWriter.java

```diff
@@ -0,0 +1,107 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink.sink.writer;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
+import org.apache.flink.api.connector.sink2.SinkWriter;
+import org.apache.flink.api.connector.sink2.StatefulSink.StatefulSinkWriter;
+import org.apache.flink.api.connector.sink2.TwoPhaseCommittingSink;
+import org.apache.flink.table.data.RowData;
+import org.apache.iceberg.flink.sink.TaskWriterFactory;
+import org.apache.iceberg.flink.sink.committer.FilesCommittable;
+import org.apache.iceberg.io.TaskWriter;
+import org.apache.iceberg.relocated.com.google.common.base.MoreObjects;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+
+public class StreamWriter
+    implements StatefulSinkWriter<RowData, StreamWriterState>,
+        SinkWriter<RowData>,
+        TwoPhaseCommittingSink.PrecommittingSinkWriter<RowData, FilesCommittable> {
+  private static final long serialVersionUID = 1L;
+
+  private final String fullTableName;
+  private final TaskWriterFactory<RowData> taskWriterFactory;
+  private transient TaskWriter<RowData> writer;
+  private transient int subTaskId;
+  private final List<FilesCommittable> writeResultsState = Lists.newArrayList();
+
+  public StreamWriter(
+      String fullTableName,
+      TaskWriterFactory<RowData> taskWriterFactory,
+      int subTaskId,
+      int numberOfParallelSubtasks) {
+    this.fullTableName = fullTableName;
+    this.subTaskId = subTaskId;
+
+    this.taskWriterFactory = taskWriterFactory;
+    // Initialize the task writer factory.
+    taskWriterFactory.initialize(numberOfParallelSubtasks, subTaskId);
+    // Initialize the task writer.
+    this.writer = taskWriterFactory.create();
+  }
+
+  @Override
+  public void write(RowData element, Context context) throws IOException, InterruptedException {
+    writer.write(element);
+  }
+
+  @Override
+  public void flush(boolean endOfInput) throws IOException {}
+
+  @Override
+  public void close() throws Exception {
+    if (writer != null) {
+      writer.close();
+      writer = null;
+    }
+  }
+
+  @Override
+  public String toString() {
+    return MoreObjects.toStringHelper(this)
+        .add("table_name", fullTableName)
+        .add("subtask_id", subTaskId)
+        .toString();
+  }
+
+  @Override
+  public Collection<FilesCommittable> prepareCommit() throws IOException {
+    writeResultsState.add(new FilesCommittable(writer.complete()));
+    this.writer = taskWriterFactory.create();
+    return writeResultsState;
+  }
+
+  @Override
+  public List<StreamWriterState> snapshotState(long checkpointId) {
+    List<StreamWriterState> state = Lists.newArrayList();
+    state.add(new StreamWriterState(Lists.newArrayList(writeResultsState)));
```

Comment:
Checkpoint individual data files as writer state can be expensive. E.g., if one checkpoint cycle produces a lot of files with many columns, the amount of metadata can be very significant.

I am also think we shouldn't checkpoint the flushed files here. What does it help?

In the current `FlinkSink` implementation, `IcebergFilesCommitter` checkpoint a staging `ManifestFile` object per checkpoint sync. Here we only need to checkpoint a single manifest file per cycle, which is very small in size. Upon successful commit, committer clears the committed manifest files. Upon commit failure, manifest files are accumulate and attempted for commit in the next checkpoint cycle. 

Checkpoint the flushed data files here doesn't bring fault tolerance in my understanding.

Reply:
`writeResultsState` contains only the metadata of the data file. We don't really put data into the state.

If we remove the state here, this part of the data may be lost if the task fails after the file is closed and before it is sent to the committer.

The more important reason is, in the v2 Sink , `CommT` is transferred between the Writer and Commiter, flink automatically processes its state. If we're going to use `Manifest` instread of `WriteResult` as `CommT`, it means need to generate it in the every writer operator, so that it can be passed to the commiter, i think that's not reasonable. If we generate it in Commiter, as the `IcebergFilesCommitter` did, because the `CommT` still is `WriteResult`, the new version of Commiter doesn't put it in the state, so there is no practical point in generating it. Given the differences between the v1 and v2 versions, I currently feel it would be better to use WriterResult directly.

Reply:
I am not saying with should use `Manifest` as `CommT`. Again, this is where I feel the v2 sink interface is not a good fit for Iceberg. In the v1 sink interface, we have `GlobalCommT` for `GlobalCommitter`. That will be the ideal interface for Iceberg sink. `WriteResult` would be `CommT` and `Manifest` will be `GlobalCommT`.

```
public interface GlobalCommitter<CommT, GlobalCommT>
```

Reply:
Here is the how that Iceberg sink works today and how I hope that it should continue to work with v2 sink interface. Writers flushed the data files and send `WriteResult` to the global committer. The global committer checkpoints the `Manifest` file for fault tolerance/exactly-once.

I understand that `WriteResult` only contains the metadata. It can still be substantial in certain use cases. E.g.,
- writer parallelism is 1,000
- Each writer writes to 100 files (e.g. bucketing or hourly partition by event time)
- table has many columns. with column-level stats (min, max, count, null etc.). it would be significant metadata bytes per data file. say there are 200 columns, and each columns stats take up 100 bytes. that is 20 KB per file.
- each checkpoint would have size of 20 KB x 100 x 1,000. that is 2 GB.
- if checkpoint interval is 5 minutes and commit can't succeed for 10 hours, we are talking about 120 checkpoints. Then Iceberg sink would accumulate 240 GB of state.

I know this is the worst-case scenario, but it is the motivation behind the current IcebergSink design where the global committer only checkpoint one `ManifestFile` per checkpoint cycle.

Reply:
To adapt to the new architecture, I've made a compromise here. Serialize the WriteResult sent by the StreamWriter into a Manifest file as CommT and deserialize the file in the Committer.